src/resolve/resolved-dns-netservice.c

```diff
@@ -0,0 +1,159 @@
+/***
+  This file is part of systemd.
+
+  Copyright 2017 Dmitry Rozhkov
+
+  systemd is free software; you can redistribute it and/or modify it
+  under the terms of the GNU Lesser General Public License as published by
+  the Free Software Foundation; either version 2.1 of the License, or
+  (at your option) any later version.
+
+  systemd is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include "conf-files.h"
+#include "conf-parser.h"
+#include "list.h"
+#include "resolved-dns-netservice.h"
+#include "resolved-dns-rr.h"
+#include "resolved-manager.h"
+#include "strv.h"
+
+const char* const netservice_dirs[] = {
+        "/etc/systemd/resolve",
+        "/run/systemd/resolve",
+        "/usr/lib/systemd/resolve",
+#ifdef HAVE_SPLIT_USR
+        "/lib/systemd/resolve",
+#endif
+    NULL
+};
+
+static void dns_netservice_free(DnsNetservice *netservice) {
+        if (!netservice)
+                return;
+
+        assert(netservice->manager->n_netservices > 0);
+        LIST_REMOVE(netservices, netservice->manager->netservices, netservice);
+        netservice->manager->n_netservices--;
+
+        dns_resource_record_unref(netservice->ptr_rr);
+        dns_resource_record_unref(netservice->srv_rr);
+        dns_resource_record_unref(netservice->txt_rr);
+
+        free(netservice->filename);
+        free(netservice->name);
+        free(netservice->type);
+        free(netservice->instance_name);
+        strv_free(netservice->txt);
```

Comment:
iirc TXT records may contain NUL bytes, no? this can't be a simply strv hence...